### PR TITLE
[SVN] Add several tests with different locking mechanism

### LINKED ISSUE #3638: Export of readonly conanfile.py results in permission denied error
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

I am using Conan 1.8.0-dev, Win10 & Python3.6.

When having an readonly `conanfile.py` file (SVN locking mechanism make them readonly), we get the follwing error message:
`ERROR: [Errno 13] Permission denied: 'C:\\...\\user\\svn25\\export\\conanfile.py'`

Because `_export_conanfile` function uses `shutil.copy2` all stat info including readonly are taken. May be add an option, to ignore readonly flag.

### PR DESCRIPTION
**Pull request by jgsogo**
Add tests related to #3748 and #3638

Changelog: Omit
Docs: Omit

- [x] built on top of PR #3871 

closes #3638 

@tags: svn

### CODE REVIEW

**Comment by jgsogo** [comment]
You are completely right about the `propset` commands, I forgot to commit the property. Now tests are executing `svn commit` after each lock (even after `svn lock`, although I think in that case it does nothing).

I've added more tests, now it should be covered every single option.

File: conans/client/cmd/export.py [hunk 1]
```diff
@@ -141,6 +142,8 @@ def _replace_scm_data_in_conanfile(conanfile_path, scm_data):
     new_text = "scm = " + ",\n          ".join(str(scm_data).split(",")) + "\n"
     content = content.replace(to_replace[0], new_text)
     content = content if not headers else ''.join(headers) + content
+    if not os.access(conanfile_path, os.W_OK):
+        os.chmod(conanfile_path, stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO)  # 0777
```
Comment (lasote) [inline conans/client/cmd/export.py]: no 777 ever, if it needs `os.W_OK` add it.

**Comment by jgsogo** [comment]
Without the lines

```python
    if not os.access(conanfile_path, os.W_OK):
        os.chmod(conanfile_path, stat.S_IRWXU)  # read, write, execute/search by owner
```

all the tests with the `propset`are failing (mac py3), for those not used to SVN, what do these tests do (with different flavors)?
 * set the property `svn:needs-lock` to `yes` in some files (`conanfile.py`): it forces all the users that want to modify a file to:
   * `lock` it (to remove the _readonly_), this command will fail if other user has already _locked_ the file.
   * modify
   * `commit`.
   * `unlock` (to allow other users to lock it)
 * run an `export` command

When using `scm` with `auto` there are two copies of the `conanfile.py` to the export sources:
 * the one always happens (the file is copied with as readonly)
 * another one after the `scm` replacement (try to override a readonly file)

----

An alternative solution that comes up to my mind is deleting the `export/conanfile.py` before saving it, just replace those files with a simple

```python
os.remove(conanfile_path)
```

**Comment by jgsogo** [comment]
Please, @lasote @climblinne, review again.

File: conans/client/cmd/export.py [hunk 2]
```diff
@@ -144,6 +145,10 @@ def _replace_scm_data_in_conanfile(conanfile_path, scm_data):
     new_text = "scm = " + ",\n          ".join(str(scm_data).split(",")) + "\n"
     content = content.replace(to_replace[0], new_text)
     content = content if not headers else ''.join(headers) + content
+
+    if not os.access(conanfile_path, os.W_OK):
+        os.chmod(conanfile_path, stat.S_IRWXU)  # read, write, execute/search by owner
+
```
Comment (jgsogo) [inline conans/client/cmd/export.py]: I would have made a `os.remove(conanfile_path)` before overwritting it, but it fails in Windows with a `PermissionDenied` error 😞

Comment (lasote) [reply conans/client/cmd/export.py]: I think I would prefer to not alter the conanfile but force the deletion only when present, something similar to the `rmdir` i the `files.py` but for a file. Encapsulated in the files module. It is almost the same but a bit less creepy.

Comment (jgsogo) [reply conans/client/cmd/export.py]: AFAIK it is ok to have the `conanfile.py` with those permissions. I've added the helper function, have a look at the last commit.

File: conans/util/files.py [hunk 3]
```diff
@@ -212,6 +212,21 @@ def rmdir(path):
...
 
+def _change_file_permissions(func, path, exc_info):
+    os.chmod(path, stat.S_IRWXU)
+    os.remove(path)
+
+
+def remove(path):
+    try:
+        assert os.path.isfile(path)
+        shutil.rmtree(path, onerror=_change_file_permissions)
```
Comment (lasote) [inline conans/util/files.py]: I don't understand why this work, `shutil.rmtree` usually fails if you don't pass a file, is the "onerror" making it happen somehow?

Comment (jgsogo) [reply conans/util/files.py]: That's it, it fails and the `onerror` function is handling it (it was taken from an example where the delete file/dir was handled by the same function).

I will change it to try-os.remove-catch and handle the permission error, it is better is this function is only going to accept a file 👍

**Comment by lasote** [comment]
Changelog and Docs tags are missing

**Comment by jgsogo** [comment]
I think we can omit both of them (I've updated the PR description)

**Pull request merged**